tiledb/api/src/context.rs

```diff
@@ -68,95 +68,110 @@ impl Context {
+            )
+        })?;
+
+        assert!(!c_json.is_null());
+        let raw = RawStatsString::Owned(c_json);
+        let json = unsafe { std::ffi::CStr::from_ptr(*raw) };
+        Ok(String::from(json.to_string_lossy()))
+    }
+
     pub fn get_config(&self) -> TileDBResult<Config> {
         let mut c_cfg: *mut ffi::tiledb_config_t = out_ptr!();
-        let res = unsafe { ffi::tiledb_ctx_get_config(*self.raw, &mut c_cfg) };
-        if res == ffi::TILEDB_OK {
-            Ok(Config {
-                raw: RawConfig::Owned(c_cfg),
-            })
-        } else {
-            Err(self.expect_last_error())
-        }
+        self.capi_call(|ctx| unsafe {
+            ffi::tiledb_ctx_get_config(ctx, &mut c_cfg)
+        })?;
+
+        Ok(Config {
+            raw: RawConfig::Owned(c_cfg),
+        })
     }
 
-    pub fn get_last_error(&self) -> Option<Error> {
+    // This is a bit awkard. We *expect* this to return an Err variant of the
+    // result, and then expect_last_error returns an error when it doesn't.
+    //
+    // I don't like it, but its localized to these two functions so continuing
+    // on and ignoring it.
+    pub fn get_last_error(&self) -> TileDBResult<()> {
         let mut c_err: *mut ffi::tiledb_error_t = out_ptr!();
-        let res =
-            unsafe { ffi::tiledb_ctx_get_last_error(*self.raw, &mut c_err) };
-        if res == ffi::TILEDB_OK && !c_err.is_null() {
-            Some(Error::from(RawError::Owned(c_err)))
+        self.capi_call(|ctx| unsafe {
+            ffi::tiledb_ctx_get_last_error(ctx, &mut c_err)
+        })?;
+
+        if !c_err.is_null() {
+            Err(Error::from(RawError::Owned(c_err)))
         } else {
-            None
+            Ok(())
         }
     }
 
     pub fn expect_last_error(&self) -> Error {
-        self.get_last_error()
-            .unwrap_or(Error::Internal(String::from(
-                "libtiledb: expected error data but found none",
-            )))
+        let result = self.get_last_error();
+        if result.is_err() {
+            return result.err().unwrap();
+        }
+
+        Error::Internal(String::from(
+            "libtiledb: expected error data but found none",
+        ))
     }
 
-    pub fn is_supported_fs(&self, fs: Filesystem) -> bool {
+    pub fn is_supported_fs(&self, fs: Filesystem) -> TileDBResult<bool> {
         let mut supported: i32 = 0;
-        let res = unsafe {
-            ffi::tiledb_ctx_is_supported_fs(
-                *self.raw,
-                fs as u32,
-                &mut supported,
-            )
-        };
-        if res == ffi::TILEDB_OK {
-            supported == 1
-        } else {
-            false
-        }
+        self.capi_call(|ctx| unsafe {
+            ffi::tiledb_ctx_is_supported_fs(ctx, fs as u32, &mut supported)
+        })?;
+
+        Ok(supported == 1)
     }
 
     pub fn set_tag(&self, key: &str, val: &str) -> TileDBResult<()> {
```

Comment:
Based on the discussion in `#core` the other day, this should be required as part of the context construction? Should we file a story for that?

Reply:
Done: https://app.shortcut.com/tiledb-inc/story/46092